### DISCUSSION TITLE
Correct way to add cmp-cmdline + cmdline-history to LazyVim

### DISCUSSION THREAD

**Discussion by moonstruckmoth** [LazyVim/LazyVim#2487] [Q&A]
I tried adding cmp-cmdline, cmp-cmdline-history using the config show below. It works on the whole but it seems to be interfering with my LSP completions...I need to press c-n/c-p to trigger the drop down. 

Without these plugins enabled, the completions appear automatically.

Since LazyVim already has nvim-cmp configured, my config seems to be interfering with some settings.

TIA,
Raj

```
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      "hrsh7th/cmp-cmdline",
      "dmitmel/cmp-cmdline-history",
    },
    event = "InsertEnter",
    config = function()
      local cmp = require("cmp")
      -- `/` cmdline setup.
      cmp.setup.cmdline("/", {
        mapping = cmp.mapping.preset.cmdline(),
        sources = {
          { name = "buffer" },
        },
      })

      -- `:` cmdline setup.
      cmp.setup.cmdline(":", {
        mapping = cmp.mapping.preset.cmdline(),
        sources = cmp.config.sources({
          {
            name = "cmdline",
            option = {
              ignore_cmds = { "Man", "!" },
            },
          },
          { name = "cmdline_history" },
          { name = "path" },
        }),
      })
    end,
  },


```

**Comment by dpetka2001** [comment]
Please take a look at #2072.

**Reply by moonstruckmoth** [reply]
My config is pretty much similar to #2072, but like I mentioned it's affecting my LSP completions (LazyVim's default completions).
I would like my completions to appear as I type instead of being triggered by c-n/c-p.

But thanks for ur response.

**Reply by dpetka2001** [reply]
With the config in #2072 it doesn't affect LSP completions in my config. You could try to use that config instead of yours and see how it behaves and if it meets your needs or not.

**Reply by dpetka2001** [reply]
[Screencast 2024-02-03 13:22:43.webm](https://github.com/LazyVim/LazyVim/assets/12776461/db110810-3b8c-4037-8b22-ee1237b8e937)

Here's a screencast of how it looks for me.

**Reply by moonstruckmoth** [reply]
Thanks for all your help @dpetka2001.

My requirement is that when I go to command line using ':', I want suggestions from cmdline, cmdline_history and path.

With the configuration I have mentioned in the original post, I get these completions but LSP doesn't get autotriggered.

I tried adding the config you suggested + similar config for cmdline-history. Here LSP works fine but I'm only getting completions for cmdline without cmdline-history. 

Here is the new config.

Thanks again for spending time on this

```
  {
    "hrsh7th/cmp-cmdline",
    event = "InsertEnter",
    config = function()
      local cmp = require("cmp")
      ---@diagnostic disable-next-line: missing-fields
      cmp.setup.cmdline(":", {
        mapping = cmp.mapping.preset.cmdline({
          ["<C-j>"] = {
            c = function(fallback)
              if cmp.visible() then
                cmp.select_next_item()
              else
                fallback()
              end
            end,
          },
        }),
        sources = cmp.config.sources({
          { name = "path" },
        }, {
          {
            name = "cmdline",
            option = {
              ignore_cmds = { "Man", "!" },
            },
          },
        }),
      })
    end,
  },

  {
    "dmitmel/cmp-cmdline-history",
    event = "InsertEnter",
    config = function()
      local cmp = require("cmp")
      cmp.setup.cmdline(":", {
        sources = {
          { name = "cmdline_history" },
        },
      })
    end,
  },

```

**Reply by dpetka2001** [reply]
Yes, the config I mentioned in the other issue didn't have `cmdline-history` because it wasn't a requirement in that thread. And from your title I thought you only had problems with `cmp-cmdline`, that's why I didn't notice the `cmdline-history` in your initial setup.

**Reply by moonstruckmoth** [reply]
I finally got everything working with the following config. I'm not sure if it's the most optimum one, but it does what I need.
Thanks everyone who spent time on this

```
   {
    "hrsh7th/cmp-cmdline",
    event = "InsertEnter",
    config = function()
      local cmp = require("cmp")
      local config = cmp.get_config()
      table.insert(config.sources, {
        name = "path",
      })
      table.insert(config.sources, {
        name = "cmdline",
        option = {
          ignore_cmds = { "Man", "!" },
        },
      })
      config.mapping = cmp.mapping.preset.cmdline()
      -- `:` cmdline setup.
      cmp.setup.cmdline(":", config)
      -- `/` cmdline setup.
      cmp.setup.cmdline("/", {
        mapping = cmp.mapping.preset.cmdline(),
        sources = {
          { name = "buffer" },
        },
      })
      cmp.setup(config)
    end,
  },

  {
    "dmitmel/cmp-cmdline-history",
    event = "InsertEnter",
    config = function()
      local cmp = require("cmp")
      local config = cmp.get_config()
      table.insert(config.sources, {
        name = "cmdline_history",
      })
      config.mapping = cmp.mapping.preset.cmdline()
      cmp.setup.cmdline(":", config)
      cmp.setup(config)
    end,
  },


```

**Comment by moonstruckmoth** [comment]
The config I shared earlier resolved my requirements
